### PR TITLE
test(ruff_python_formatter): Run all Black tests

### PR DESCRIPTION
**Pull request by MichaReiser**
This PR changes the testing infrastructure to run all black tests and:

* Pass if Ruff and Black generate the same formatting
* Fail and write a markdown snapshot that shows the input code, the differences between Black and Ruff, Ruffs output, and Blacks output

This is achieved by introducing a new `fixture` macro (open to better name suggestions) that "duplicates" the attributed test for every file that matches the specified glob pattern. Creating a new test for each file over having a test that iterates over all files has the advantage that you can run a single test, and that test failures indicate which case is failing. 

The `fixture` macro also makes it straightforward to e.g. setup our own spec tests that test very specific formatting by creating a new folder and use insta to assert the formatted output.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,5 @@
 [workspace]
 members = ["crates/*"]
-default-members = ["crates/ruff", "crates/ruff_cli"]
```
Comment (MichaReiser) [inline Cargo.toml:3]: TODO: Move to its own PR

File: crates/ruff_python_formatter/src/snapshots/ruff_python_formatter__tests__black_test__attribute_access_on_number_literals_py.snap [hunk 2]
```diff
@@ -0,0 +1,148 @@
+---
+source: crates/ruff_python_formatter/src/lib.rs
+expression: snapshot
+input_file: crates/ruff_python_formatter/resources/test/fixtures/black/simple_cases/attribute_access_on_number_literals.py
+---
+## Input
+
+
+```py
+x = 123456789 .bit_count()
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/snapshots/ruff_python_formatter__tests__black_test__attribute_access_on_number_literals_py.snap:10]: You can override the filetype for .py.snap in your JetBrains IDE to get markdown rendering

image

File: crates/ruff_testing_macros/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+                pattern,
+                pattern_span,
+            }),
+        }
+    }
+}
+
+#[proc_macro_attribute]
+pub fn fixture(attribute: TokenStream, item: TokenStream) -> TokenStream {
+    let test_fn = parse_macro_input!(item as ItemFn);
```
Comment (MichaReiser) [inline crates/ruff_testing_macros/src/lib.rs:79]: This is where all the fun of the new macro is happening

**Comment by MichaReiser** [comment]
Hmm. The windows tests are all failing now because the markdown file is written using `writeln` that always inserts `\n` but the checked out snapshots use `\cr\n` because we use no `.gitattributes` file. 

I tried to add a `.gitattributes` file that ensures that all platforms use the same line endings when checking out the repository. 

```
* text=auto eol=auto

crates/ruff/resources/test/fixtures/isort/line_ending_crlf.py text eol=crlf
crates/ruff/resources/test/fixtures/pycodestyle/W605_1.py text eol=crlf

ruff.schema.json linguist-generated=true text=auto eol=lf
```

But now many of the linter tests are failing which seems a bit frightening. Does it mean that ruff doesn't correctly handle the case where a project uses the non-default line ending of the operating system?

```
━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Differences ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot file: crates\ruff\src\rules\flake8_comprehensions\snapshots\ruff__rules__flake8_comprehensions__tests__C400_C400.py.snap
Snapshot: C400_C400.py
Source: C:\Users\Micha\git\ruff:39
───────────────────────────────────────────────────────────────────────────────
Expression: diagnostics
───────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬──────────────────────────────────────────────────────────────────
   24    24 │     row: 4
   25    25 │     column: 1
   26    26 │   fix:
   27    27 │     content:
   28       │-      - "["
   29       │-      - "    x for x in range(3)"
   30       │-      - "]"
         28 │+      - "[\n    x for x in range(3)\n]"
   31    29 │     location:
   32    30 │       row: 2
   33    31 │       column: 4
   34    32 │     end_location:
────────────┴──────────────────────────────────────────────────────────────────
```

Any ideas where I should start looking? Is this in the parser or does the `stylist` infer the wrong line endings (why are new lines embedded in the string literal)?

**Comment by andersk** [comment]
#2033 has some relevant context, especially the `add_redaction` call in our `assert_yaml_snapshot` wrapper.

**Comment by MichaReiser** [comment]
> #2033 has some relevant context, especially the `add_redaction` call in our `assert_yaml_snapshot` wrapper.

Ohh thanks! It would have taken me forever to find this

**Comment by MichaReiser** [comment]
Depends on #3005 to fix the windows build (or we have to handle the line ending differences as part of this PR)

File: crates/ruff_python_formatter/src/lib.rs [hunk 4]
```diff
@@ -57,70 +55,106 @@ pub fn fmt(contents: &str) -> Result<Formatted<ASTFormatContext>> {
...
+            "*fmtskip8.py",
+            "*function2.py",
+            "*one_element_subscript.py",
+            "*remove_except_parens.py",
+            "*remove_parens.py",
+            "*remove_newline_after_code_block_open.py",
+            "*slices.py",
+            "*string_prefixes.py",
+            "*tricky_unicode_symbols.py",
+            "*tupleassign.py"
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/lib.rs:87]: There are now more tests that panic than when I initially created this PR. I think it would be good for us to fix the panics first to get better visibility of our changes (and avoid having any new tests that cause panics)

File: crates/ruff_python_formatter/src/lib.rs [hunk 5]
```diff
@@ -57,70 +55,98 @@ pub fn fmt(contents: &str) -> Result<Formatted<ASTFormatContext>> {
...
+            let mut snapshot = String::new();
+            write!(snapshot, "{}", Header::new("Input"))?;
+            write!(snapshot, "{}", CodeFrame::new("py", &content))?;
+
+            write!(snapshot, "{}", Header::new("Black Differences"))?;
+
+            let diff = TextDiff::from_lines(expected_output.as_str(), formatted_code)
+                .unified_diff()
+                .header("Black", "Ruff")
+                .to_string();
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/lib.rs:135]: That's correct. I found it useful when working on rome formatter to have all information in a single file, so that you don't need to search for the expected output file.

File: crates/ruff_python_formatter/src/lib.rs [hunk 6]
```diff
@@ -150,4 +176,41 @@ mod tests {
...
+        title: &'a str,
+    }
+
+    impl<'a> Header<'a> {
+        fn new(title: &'a str) -> Self {
+            Self { title }
+        }
+    }
+
+    impl std::fmt::Display for Header<'_> {
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/lib.rs:190]: Thanks. It may be familiar from working on the formatter ;)

File: crates/ruff_testing_macros/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+[package]
+name = "ruff_testing_macros"
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_testing_macros/Cargo.toml:2]: In my view it mainly comes down if you prefer having fewer crates or more fine granular crates. 

My preference is to have more fine granular crates because:

* Smaller crates mean there are fewer changes that require a re-compile of its code
* You can be explicit about dependencies
  * the testing macros don't need to depend on itertools
  * The test macros can be a `dev` dependency, ensuring that its functionality isn't used in production code
* I find smaller crates help navigate the code base and specific names tell you if this crate is the right crate for your code addition

The main downside is that we end up with more crates, resulting in a longer `crates` list. This doesn't seem to be a problem today.

File: crates/ruff_python_formatter/src/snapshots/ruff_python_formatter__tests__black_test__beginning_backslash_py.snap [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+## Black Output
+
+```py
+print("hello, world")
+
+
+```
+
+
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/snapshots/ruff_python_formatter__tests__black_test__beginning_backslash_py.snap:46]: I would have thought so but that doesn't seem to be the case. Using `trim_end()`  on the generated content doesn't change anything and the two trailing new lines are present in other snapshot tests. 

https://github.com/MichaReiser/ruff/blob/b306a17cdc85c11f9cc8e1e27a5d88959180ac78/crates/ruff/src/rules/isort/snapshots/ruff__rules__isort__tests__lines_after_imports_lines_after_imports_nothing_after.py.snap#L1-L0
https://github.com/charliermarsh/ruff/blob/b306a17cdc85c11f9cc8e1e27a5d88959180ac78/crates/ruff/src/rules/isort/snapshots/ruff__rules__isort__tests__lines_after_imports_lines_after_imports_nothing_after.py.snap#L1-L0

File: crates/ruff_python_formatter/src/snapshots/ruff_python_formatter__tests__black_test__beginning_backslash_py.snap [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+```py
+print("hello, world")
+```
+
+## Black Output
+
+```py
+print("hello, world")
+
+
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/snapshots/ruff_python_formatter__tests__black_test__beginning_backslash_py.snap]: Done. I added an editorconfig file that sets `final_new_line = false` for the expect files.
